webapp/src/views/projects/languages/MachineTranslation/MachineTranslation.tsx

```diff
@@ -0,0 +1,317 @@
+          queryClient.setQueriesData(
+            matchUrlPrefix(
+              '/v2/projects/{projectId}/machine-translation-service-settings'
+            ),
+            {
+              _version: Math.random(), // force new instance
+              ...data,
+            }
+          );
+        }
+      })
+      .catch(() => {
+        setFormInstance((i) => i + 1);
+      });
+    lastUpdateRef.current = promise;
+  };
+
+  const baseSetting = settings.data?._embedded?.languageConfigs?.find(
+    (item) => item.targetLanguageId === null
+  );
+
+  const providers = Object.entries(config.machineTranslationServices.services)
+    .map(([service, value]) => (value.enabled ? service : undefined))
+    .filter(Boolean) as string[];
+
+  const gridTemplateColumns = `1fr ${providers
+    .map(() => 'auto')
+    .join(' ')} auto`;
+
+  const langDefaults: Record<
+    string,
+    { enabled: string[]; primary?: string | null }
+  > = {};
+
+  languages.data?._embedded?.languages?.forEach((lang) => {
+    const config = settings.data?._embedded?.languageConfigs?.find(
+      (langSettings) => langSettings.targetLanguageId === lang.id
+    );
+
+    langDefaults[lang.tag] = {
+      enabled: config?.enabledServices || baseSetting?.enabledServices || [],
+      primary: config ? config.primaryService || 'none' : 'default',
+    };
+  });
+
+  const initialValues = {
+    default: {
+      enabled: baseSetting?.enabledServices,
+      primary: baseSetting?.primaryService || 'none',
+    },
+    languages: langDefaults,
+  };
+
+  const isFetching =
+    settings.isFetching || languages.isFetching || creditBalance.isFetching;
+
+  useGlobalLoading(isFetching);
+
+  const isUpdating = updateSettings.isLoading;
+
+  const formatLangSettings = (
+    lang: string | null,
+    enabled: string[],
+    primary: string | null | undefined
+  ): MachineTranslationLanguagePropsDto | null => {
+    const targetLanguageId = lang
+      ? languages.data?._embedded?.languages?.find((l) => l.tag === lang)?.id
+      : null;
+
+    if (primary === 'default') {
+      return null;
+    }
+
+    return {
+      targetLanguageId: targetLanguageId as any,
+      primaryService: (primary === 'none' ? null : primary) as any,
+      enabledServices: enabled as any,
+    };
+  };
+
+  useEffect(() => {
+    formRef.current?.resetForm();
+  }, [settings.data]);
+
+  useEffect(() => {
+    if (Number(settings.data?._embedded?.languageConfigs?.length) > 1) {
+      setExpanded(true);
+    }
+  }, [settings.data]);
+
+  return (
+    <>
+      {settings.data && languages.data && creditBalance.data && (
+        <div className={classes.container}>
+          <div className={tableClasses.table} style={{ gridTemplateColumns }}>
+            <Formik
+              key={formInstance}
+              initialValues={initialValues}
+              enableReinitialize={true}
+              onSubmit={() => {}}
+              validate={(values) => {
```

Comment:
Maybe extract the validate fn out of JSX?

Reply:
It would also be clear, that you are hacking the validate function for submission.

Reply:
👍